core/dispatch.c

```diff
@@ -874,6 +874,17 @@ dispatch_enter_dynamorio(dcontext_t *dcontext)
         if (get_at_syscall(dcontext))
             handle_post_system_call(dcontext);
 
+#ifdef X86
+        /* If the next basic block starts at a debug register value,
+         * we fire a single step exception before getting to the basic block. */
```

Comment:
Is that how your existing impl works w/ a mid-bb debug reg too?  The client's instru does not execute for that instr before the fault?